### ISSUE TITLE
Include digital signatures on project packages

### ISSUE THREAD

**Issue by Gormartsen** [backdrop/backdrop-issues#1992]
I propose to implement MD5 checksum for each packaged project.

It can be  provided as a part of the title.

We need to use it via project installer to make sure that package is not hacked and replaced.

---

Advocates: @jlfranklin @quicksketch 

---

Weekly update?

---

~Proof of concept by quicksketch: https://github.com/backdrop-ops/backdropcms.org/pull/456~
Initial core PR by @jlfranklin: https://github.com/backdrop/backdrop/pull/2449

Other related issues that also need to be done:
- [x] Project Module API improvements: https://github.com/backdrop-contrib/project/issues/25
- [ ] Project Module Code Sign Support: https://github.com/backdrop-contrib/project/issues/29
- [ ] Update Installer Module: https://github.com/backdrop/backdrop-issues/issues/3714

**Comment by Gormartsen** [comment]
Nice example is here:
https://github.com/syncthing/syncthing/releases/

**Comment by klonos** [comment]
👍 Security++

**Comment by jenlampton** [comment]
👍

**Comment by jlfranklin** [comment]
Ooh.  SyncThing also GPG signs their releases.  

MD5, while good for a lot of things, has been weakened enough that it shouldn't be used for making sure a package hasn't been intentionally hacked.

I like the idea, but I suggest a stronger hash algorithm... SHA256, maybe.  Even better, GPG sign them or encourage module maintainers to do so.

**Comment by klonos** [comment]
> ...GPG sign them or encourage module maintainers to do so.

If our packager can do that for them, then we should do that for them :) ...basically the plan that @quicksketch has laid out.

**Comment by quicksketch** [comment]
I've been making progress on this functionality. Adding MD5 capability to the Project Release module is fairly trivial. So that shouldn't be difficult at all.

Digital signing also doesn't appear to be too difficult, though we need to make some architecture decisions.

Using the Sodium Compat library at https://github.com/paragonie/sodium_compat, the basics of signing are provided in the README.md file:

```php
<?php
require_once "/path/to/sodium_compat/autoload.php";

$alice_kp = sodium_crypto_sign_keypair();
$alice_sk = sodium_crypto_sign_secretkey($alice_kp);
$alice_pk = sodium_crypto_sign_publickey($alice_kp);

$message = 'This is a test message.';
$signature = sodium_crypto_sign_detached($message, $alice_sk);
if (sodium_crypto_sign_verify_detached($signature, $message, $alice_pk)) {
    echo 'OK', PHP_EOL;
} else {
    throw new Exception('Invalid signature');
}
```

Applying this to Backdrop would mean the following:

- We would generate a key-pair per the above code.
- The private key would be stored in a locked-down location on backdropcms.org server.
- The public key would be bundled directly into Backdrop core, probably under `/core/misc/keys` or in `/core/modules/installer`.
- When the packager runs on backdropcms.org, it would md5 hash the archive file, and store this md5 value in the `project_release` table.
- After doing the md5, the packager would also create a digital signature, using `$signature = sodium_crypto_sign_detached($message, $secret_key)`. The "message" would be the md5 value of the archive. It would also store this signature in the `project_release` table.
- Both the MD5 and the signature will be added to the project XML feed, such as the one at https://updates.backdropcms.org/release-history/backdrop/1.x
- When the Installer module downloads a new module or release, it will first validate the MD5 of the archive. Then it will take the signature value, and decode it with `sodium_crypto_sign_verify_detached($signature, $message, $alice_pk)`. Where the $message is the MD5 string again, and the public key is the one bundled with Backdrop core.
- Now we know that the package both has not been tampered with, and it was packaged exclusively through the private key that only exists on the BackdropCMS.org server.

Some additional thoughts:

- We should probably create not just one, but two key pairs. The first private key is stored on backdropcms.org. The second private key is a backup we would store only in an offline location. Both public keys would be bundled with Backdrop core. The backup would be used in the event the main private key becomes compromised, in which case we would regenerate all the signatures using the new key.
- `sodium_crypto_sign_detached()` creates a string that is made up of random characters, so it doesn't seem to have any real sane representation when viewed in a text file. Here's what a public key looks like:
```
çhú=!M/Åv¼{D½˜—æ>þœ¯Èfçüð·N-ðµ˜B-ÀÂÙæCØs(Œé‘4/CPÜ¡
Ä†þLµ˜B-ÀÂÙæCØs(Œé‘4/CPÜ¡
Ä†þL
```
- As this string is likely to be interpretted incorrectly when transferred as text, we should encode it in some way before we store it. The [AirShip CMS](https://github.com/paragonie/airship/) made by Paragonie provides some example as to how to do this, where they use base64 URL-safe encoding before storing and writing keys to disk.

In `\ParagonIE\Halite\Asymmetric\Crypto::sign`, the code to do the actual signing is:
```php
    public static function sign(
        string $message,
        SignatureSecretKey $privateKey,
        $encoding = Halite::ENCODE_BASE64URLSAFE
    ): string {
        $signed = \sodium_crypto_sign_detached(
            $message,
            $privateKey->getRawKeyMaterial()
        );
        $encoder = Halite::chooseEncoder($encoding);
        if ($encoder) {
            return (string) $encoder($signed);
        }
        return (string) $signed;
    }
```

Where $message is an MD5 string, and $encoding is `"base64urlsafe"`. They provide the option to use a different encoding, but I think for practical purposes, we can just assume a Base64 encoding in the BackdropCMS.org packager. I don't think we need to worry about making the encoded version URL-safe, but Backdrop does include a [`backdrop_base64_encode()`]( https://api.backdropcms.org/api/backdrop/core%21includes%21bootstrap.inc/function/backdrop_base64_encode/1) function for URL-safe strings, but it does not have an equivalent function for decoding them.

**Comment by docwilmot** [comment]
Sounds good.

**Comment by jlfranklin** [comment]
Sounds like you're reinventing SSL signing, minus the identity features.  

Where you have an offline key, SSL would use a CA key and cert.  (It's easy enough to create a CA cert using OpenSSL.)  The CA's key (private part) is kept offline, the cert itself (public part) is baked into Backdrop.  A signing key is generated and that key/cert lives on backdropcms.org for signing packages.  (Actually, it should live in a more secure signing server that has limited access to the internet, and only listens for signing requests from whitelisted sites like backdropcms.org.)

The signing cert can be recreated if (when) it is compromised, and the individual package certs can carry the identity of the package and author(?) or signer(?).  The date in the signed package's cert should match the release date of the module, and if it isn't refreshed in _n_ years when the cert expires, the module should be considered abandoned.

Separate signing certs can be generated based on characteristics of the module.  For example, regular modules would get signed by the main cert, sandbox modules would be signed with a lower-grade cert.

Using a certificate revocation list, modules with (>= serious) security releases have their signing certs revoked, as will modules that are considered abandoned before their cert expires.

There will need to be some work done to verify the CA cert is the one that signed the package, but [that code exists](http://php.net/manual/en/function.openssl-verify.php#98282).

The Backdrop project could then sell signing certs to major web shops or (cough) forks of the project as a way to generate some revenue.

**Comment by quicksketch** [comment]
> Sounds like you're reinventing SSL signing, minus the identity features.

Well, I'm not reinventing SSL signing, I'm just using libsodium key-pair signing instead of SSL certificate signing.

You bring up a good point though, we might be able use OpenSSL signed certificates instead of Sodium signing. I could see benefits and downsides: The existing Sodium functions seem to be made for this exact purpose, they're fairly trivial to implement, are (supposedly) quite secure, and are built into newer versions of PHP. On the downside, this technology is relatively new, and requires an backwards compatibility shim for older PHP versions.

OpenSSL is available for all versions of PHP, but it requires that OpenSSL be installed on the host computer and PHP configured to use it correctly. I would expect that having PHP correctly configured and `--with-openssl` compilation flag might turn into a common installation problem.

I'm also not clear on how an OpenSSL implementation would work. @jlfranklin could you attempt to stub out code you would expect to validate the signature of a downloaded package?

> The Backdrop project could then sell signing certs to major web shops

They would only need signing certs if they were redistributing their modules through another means than backdropcms.org, because the packager will already do the signing for everything in the contrib group. At this point we're just looking to make sure auto-updates come from a valid authority (backdropcms.org), so this validation only occurs when downloading a module, we wouldn't check anything for modules that already exist on disk (i.e. custom modules).

Overall on the question of OpenSSL vs Sodium (or any other option), I'm fine with any of them as long as it's secure and relatively simple. I think we should choose whichever solution is going to be the least hassle for inexperienced end-users.

**Comment by jlfranklin** [comment]
>I'm also not clear on how an OpenSSL implementation would work. @jlfranklin could you attempt to stub out code you would expect to validate the signature of a downloaded package?

I'd be happy to.  Give me through the weekend.  I'm not sure how much free time I have this week.

**Comment by quicksketch** [comment]
I haven't tested all the pieces together, but I've put up a proof of concept against the backdropcms.org repository: https://github.com/backdrop-ops/backdropcms.org/pull/456

For the time being I've just written it to use the sodium functions (though I have not included the sodium library itself yet). So far I've only tested creating a release and verifying that the MD5 is added correctly and that the MD5 + signature columns are added to the project_release table. Alternatively, we could put the signature as an attachment (asset) on the GitHub release, but it will make invalidating/regenerating signatures much easier if they're all stored centrally on backdropcms.org.

**Comment by jenlampton** [comment]
My vote would be for using Sodium because 1) I like things that are easy, for learnability reasons. and 2) the shim may be useful for servers that don't have the Sodium module compiled with PHP, even after version 7.2, and 3) since we've already started going down this road this approach may also be the shortest route to the finish line. .02 ;)

**Comment by jlfranklin** [comment]
I've worked through enough of the coding that I think I have a workable architecture for SSL-based signatures on modules.  This post is pretty dense and technical.  I don't recommend starting unless you have some time and spare cycles to devote to it.

Before we start, I want to emphasize that SSL is about _identity_, not _encryption_.  Encryption is a _prerequisite_ for SSL, not a benefit.  The same encryption algorithms used by SSL are also used by SSH.  You could use an SSH tunnel to connect to a website securely, but you would have no assurance you're talking to the right web site.  SSL provides the identity mechanism to know _who_ you're talking to.

Each SSL cert is a document that says, "I [insert signer's name here], hereby certify [insert subject's name here] is the owner of the private key matching the public key in this cert," and a digital signature that can be verified with the signer's public key.  The cert contains the Common Name or CN of both the _subject's name_ and the _signer's name_.  (If you recognize CN from LDAP, it's because SSL and LDAP both derive from X.500 and share a lot of the same terminology.)

SSL certs define a one-to-one, "A signs B" relationship.  If you need multiple entities to sign a key, you need multiple certs.

When I talk about creating certs below, I implicitly include the creation of a public-private key pair for the cert.

With all that in mind, let's begin.

As an overview, below is the certificate chain that I envision, with the CN of each cert in parenthesis. 

Root CA Cert (Backdrop Codesign CA) signs
Class _n_ Signing Cert (Backdrop Codesign Intermediate Class _n_) signs
Developer Signing Cert (Edward Xample, Developer) signs
Module Signing Cert (My Module, v1.0) creates
signature block (encrypted digest of some hash).

The Root CA Cert, Backdrop Codesign Intermediate cert, and any other CA certs are kept in `core/misc/certs`.  The Root CA's private key should be kept offline in a secure location, such as a bank deposit box.  The Codesign private key is on a hardend server that listens for signing requests from Backdrop.org and other whitelisted servers.

The developer applies for his signing cert from the Backdrop Project.  For the core code and core modules, the developer cert will have something like "Backdrop Core Team" as the CN, and be managed by the release team.

Each module has a `MANIFEST.txt` file with the file names of all module files, including submodules.  I considered using the module's `.info` file to store the manifest, but decided against it as module packages that include submodules would have multiple info files, and I thought the signature should be for the full module package. If the `MANIFEST.txt` isn't created by the author, it should be created by the signing process.

Now that the developer has his own signing cert, we're ready to sign a module.

The signing process creates a module cert signed by the developer's cert.  The name of the module and its version is the CN, and the `subjectAlternativeName` has the machine names and versions of the modules and submodules.  (E.g., my_module-1.0, my_submodule-1.0.)  The `subjectAlternativeName` allows Backdrop to easily match the modules to the signatures.

The signing process reads the manifest file and concatenates the contents of each file named in order, finally adding the manifest itself to the end.  This is the "data" that will be signed.

The signing process then uses the private key associated with the module cert to sign the module "data".  The signature -- a hash of the "data", encrypted with the signer's private key -- is written to a `codesign.pem` file in a format that saves the hashing algorithm (e.g. sha256) and the key ID of the module's signing key.  The module's cert, the developer's cert, and any other intermediate certs are added to the `codesign.pem` file.

When Backdrop attempts to verify the signature, it reads the `codesign.pem` file to get the signature block and the cert with the public key needed to verify the signature.  If the signature checks out, the verification process follows the cert chain back until it finds one of the CA certs (success) or the chain is broken (failure.)

Multiple signatures can be in the `codesign.pem` file.  A second signature from the "Backdrop Module Publishing Service" can be present in `codesign.pem`, with the Publishing Service replacing the developer's cert in the chain.

Finally, a Codesign module can add signing information to the `admin/modules` page or its own `admin/reports/codesign` page, and a cron hook to verify the code from time to time.

That's it.

So far, it's about 400 lines of PHP.  With proper docs and tests, I expect it to be around 1,000 lines total.  I've run into a couple dead ends, because some of the OpenSSL functions were introduced later than I thought, some as recently as 7.2, and I wanted to make sure we had a solution that would work with PHP 5.x. 

Obviously, the code will need some thorough vetting, but I expect the hardest part will be the infrastructure to support it.

**Comment by jlfranklin** [comment]
I would have loved to get this into 1.12, but there is no way that is going to happen.  Here is a quick update on my progress.

To get this to work requires changes in at multiple areas:

- A new core `codesign` module to handle the crypto.
- Changes to the module installer to use `codesign` to verify the downloaded modules.
- Changes to the Project module (specifically `project_release`) to allow other modules to add data to a release.
- An additional `project_codesign` module to generate and add the signatures to module releases.

Changes to Project are already done.  (See backdrop-contrib/project#25.)

The `codesign` module is mostly done, but I keep refactoring major pieces of it as I develop other parts of the system.  (Track https://github.com/sd-backdrop/backdrop/commits/BD-1992 for progress.)  The current `codesign` module is crypto-engine agnostic, allowing contrib modules to handle the signing and key management for OpenSSL, GnuPG, Sodium, or whatever the new hotness is next year.

I'm writing two contrib modules to ensure `codesign` is crypto-engine agnostic, one for OpenSSL, one for GnuPG.  They are evolving with `codesign`.

The `project_codesign` module is my current focus.  I'm debating adding it as a submodule to Project or creating a separate contrib module.

And, of course, there will need to be a suite of tests written for all the different parts.

**Comment by klonos** [comment]
Thanks for all your hard work and energy/time spent on this @jlfranklin 👍

Do we absolutely need to have a separate module for this? When we merged diff module in core (for use in the config diff in `/admin/config/development/configuration`), we have done that as an include file. I mean, is this something that we will need to turn on/off?

**Comment by jlfranklin** [comment]
It implements `hook_menu()` and will likely have some of its own configuration and tables by the time it's done.  Codesign is not required for a site to function properly, so it _could_ be turned off.

**Comment by klonos** [comment]
Thanks for taking the time to respond @jlfranklin ...have not gotten my head around everything involved yet, but this helps.

**Comment by jlfranklin** [comment]
See also https://github.com/sd-backdrop/codesign_gnupg as the first functional signing module.

**Comment by quicksketch** [comment]
Awesome @jlfranklin! I haven't checked in on this in a long time! It's great to see progress here.

> I'm debating adding it as a submodule to Project or creating a separate contrib module.

Perhaps start separate, but if we use this on backdropcms.org we'll want it directly in project module. The situations where project module is _not_ part of backdropcms.org are our secondary use-case, and it's already hard enough to manage with project module being separate from the main backdropcms.org repository.

**Comment by herbdool** [comment]
Can I quibble about the name? I got confused by "codesign" and others will too. I first read it as "co-design". I suggest making it "code_sign" so it's clearer.

**Comment by klonos** [comment]
Makes perfect sense @herbdool 👍

**Comment by jlfranklin** [comment]
> I suggest making it "code_sign" so it's clearer.

No objection.

**Comment by jlfranklin** [comment]
This has come far enough to generate code signatures, and therefore far enough to get some other people looking at it.  This PR by itself is testable, but does not achieve the goal of the ticket.  That requires some patches to the Project module (see backdrop-contrib/project#25 and backdrop-contrib/project#29) and adding the [sd-backdrop/code_sign_gnupg](https://github.com/sd-backdrop/code_sign_gnupg) or [sd-backdrop/code_sign_openssl](https://github.com/sd-backdrop/code_sign_openssl) contrib modules.

Don't look at those links, yet.  Let's go through this PR, first.

The PR adds a `core/modules/code_sign` module that, despite its name, is able to sign any data that can be passed in as a string to the signing function.  The `code_sign` module cannot sign anything by itself, it requires a signing module to do the actual signing and any key management.

For demonstration purposes and to run the core Simpletest suite, a `code_sign_hash` module is included.  It returns a simple hash of the data, using any hash supported by the PHP `hash()` function.  This is good enough for testing, but not real world module signing.

* Build a test site with PR#2449, linked above.
* Enable only the Code Sign module.

There are two new menu items that will appear in the site.  One at Configuration => System => Code Sign, the other at Reports => Code Sign.   The one under Configuration has just a dummy page where other modules can add their own config tabs.  The report page notes no signing modules are installed and invites you to install the Hashing Code Sign module.

* Enable the Hashing Code Sign module.

Now the report page lists a profile and a new Hash Signing tab is under the Configuration => System => Code Sign section where hashing profiles can be managed.

Externally, that's it.  The rest of the PR provides an interface for signing with a selected "profile" provided by a signing module, and verifying the signatures are valid.  For that, we'll need a module that uses it.

**Comment by jlfranklin** [comment]
Have you done everything in the last comment?  Feeling pretty good about it?  Good, so am I.  Let's move on to the Project module.

Here's a little background on Project, in case you haven't used it before.  The Project module manages nodes for Projects and Project Releases.  (Read: contrib modules and their releases.)  This is used on b.o to create the searchable library of contrib modules that have full releases.  Through `hook_cron`, it creates XML files with the metadata for the each project and for each release of each project.  Backdrop core downloads these XML files to generate the "Available Updates" report on everybody's sites and the Project Installer module in core downloads uses them for downloading the list of available modules and their latest versions, so you can install and upgrade modules without FTP access.

To add signatures to the XML, there are two patches to the Project module that need to be applied (but don't do it yet):

- Issue backdrop-contrib/project#25 changes how the XML is generated in Project.
- Issue backdrop-contrib/project#29 adds a `project_code_sign` module to optionally sign modules and put the signatures in the XML.

Before the patches, Project uses a lot of string concatenation to generate the XML.  This makes it difficult to add new elements into the XML, like the signature blocks.  The first patch above creates a nested array with all the metadata, passes the array to one of three new alter hooks via `backdrop_alter()` (depending on the code path), and then uses `format_xml_elements()` to generate the XML itself.

The second patch adds the Project Code Sign Integration module and adds a "Regenerate Release XML" page under Configuration => Project.  The new module generates signature blocks for the modules and inserts them in the XML via the new alter hooks introduced above.  It also saves the signatures into a database table so that regenerating the XML doesn't mean resigning everything.

The patch in backdrop-contrib/project#25 is included in the backdrop-contrib/project#30 PR.  You can pull just this PR and get both at once  (Which is why I told you not to apply them both above.)

OK, now do this:

* Checkout the `1.x-1.x` branch of Project into the modules directory
* Apply the PR backdrop-contrib/project#30
* Enable "Project Code Sign Integration" and any dependencies it wants.
* Create a Project node for testing.
* Create two Project Release nodes for the Project. The "download link" has to point to a real file that your test web site can reach. The rest of those three nodes can be test data.

We have everything we need to start signing modules.  The Project Code Sign Integration module adds a tab at Configuration => Project => Codesign to manage *how* modules are signed.  You should see the Default hash profile listed there.  If you go back over to the Hash Settings tab from the previous comment, you can add new Hash profiles and they'll all appear here.  Let's do that.

* Go to Configuration => System => Code Sign => Hash Signing
* Add a couple new Hash Signature Profiles:
  * Weak with MD2
  * Strong with SHA512
* Go to Configuration => Project => Codesign

You should see all three Hash signing profiles now: Default, Weak, and Strong.  

* Check all of the available profiles and save.
* Click the "Regenerate Release XML" tab.

There is an checkbox here to "Regenerate signatures."  All it does is delete everything from the signature cache table so the XML regeneration process will have to create new signatures.

* Click the "Regenerate Release XML" button and wait for the batch to complete.
* Look in your `files/release-history` directory. 

There should be a directory for the project node you created above, and two XML files.  Open up either one in your favorite text editor and you should see the following in each `<release>` block:

```xml
<signatures> <signature> <crypto_engine>hash</crypto_engine>
 <profile_id>default</profile_id>
 <signature_block>sha256:eba15d583a8b5ba4f7e1cfe39cf5e6b175b1da7fd38aa33003b981eef0702aed</signature_block>
</signature>
 <signature> <crypto_engine>hash</crypto_engine>
 <profile_id>strong</profile_id>
 <signature_block>sha512:8b641585bbdd9281d8af4564375002fadb2c2d2e85e425801b2e7783914b296fce8967f62ba43aece9b5686c29a89c756fa8a0b5ed89406923cedd5c36a8d373</signature_block>
</signature>
 <signature> <crypto_engine>hash</crypto_engine>
 <profile_id>weak</profile_id>
 <signature_block>md2:d62b90d16469c19374ed468570d227dd</signature_block>
</signature>
</signatures>
```

Congratulations.  You've just signed your first module.  (Or whatever test data you used.)

**Comment by jlfranklin** [comment]
> Perhaps start separate, but if we use this on backdropcms.org we'll want it directly in project module.

The `project_code_sign` module got added as a sub-module, mostly for patching convenience.  Extracting it as its own contrib module is trivial, and there are no technical blockers to doing so.

**Comment by jlfranklin** [comment]
After two long comments, we have hashes, but no signatures in the XML.  To do so, we need to install a module that does some kind of PKI-based signing.  There are two additional modules that can provide this.

* [Code Sign GnuPG](https://github.com/sd-backdrop/code_sign_gnupg)
* [Code Sign OpenSSL](https://github.com/sd-backdrop/code_sign_openssl)

Currently, the GnuPG module is a bit easier to use, so let's start with that one.

* Checkout the 1.x-1.x branch of Code Sign GnuPG into the modules directory.

You'll need to install gnupg into your system.  If you're running Linux, this is as easy as installing any other package.  For some popular distros, the following commands will install GnuPG:

* Debian or Ubuntu: `apt-get install gnupg`
* Fedora: `dnf install gnupg`
* Arch: `pacman -S gnupg`

What we're about to do here is a basic setup.  Going into the details of GnuPG features, the web-of-trust, and other details is outside the scope of this thread.  This is also not the recommended way, because the key will lack a passphrase protecting it.

First, we'll need to create a key.  The command `gpg --gen-key` will ask for a real name and an email address, and a passphrase.  It will print out other stuff, but what you see below is the important part.  When it asks for a passphrase, just hit Enter.  GnuPG will advice against skipping the passphrase, do it anyway.  (We will support passphrases when the module is finished.)

```
$ gpg --gen-key
Real name: Backdrop Test Signing Key
Email address: bdtest@example.com
You selected this USER-ID:
    "Backdrop Test Signing Key <bdtest@example.com>"
```

The key is created and in your keychain.  Now, we need to export it to a file so we can upload it to the website.

```
$ gpg --export-secret-keys > bdtest.gpg
```

If you didn't give it a passphrase above, it shouldn't ask for a passphrase now.  If all goes well, there will be a `bdtest.gpg` file in the current directory.

* Go to Configuration => System => Code Sign => GnuPG => Import key
* Upload the bdtest.gpg file.
* Click the GnuPG => Settings subtab.

There should be one familiar looking option for the Default signing key.  In Configuration => Project => Code Sign, it should show up as a signing option.  If there isn't, then the web server probably can't write to its `$HOME/.gnupg` directory.  It should look like this. (Replace www-data with the user name the web server runs as.)

```
$ ls -ld ~www-data/.gnupg
drwx------ 2 www-data root 4096 Mar 14 03:26 /var/www/.gnupg
```

* Go to Configuration => Project => Code Sign
* Enable the GnuPG signing profile and save.
* Regenerate Release XML, just like we did in the last long comment.

If the crypto gods are smiling on you, then your module's XML should include something like this:

```
<signature> <crypto_engine>gnupg</crypto_engine>
 <profile_id>74D1063DAB2EE638AEEF31423B30EC9640CF7EA0</profile_id>
 <signature_block>-----BEGIN PGP SIGNATURE-----

iQGzBAABCAAdFiEEdNEGPasu5jiu7zFCOzDslkDPfqAFAlyKAuMACgkQOzDslkDP
fqDXwwv/SREGkMzo1Iup5r8VoH5U3SWhp69Bwf5Nfq1y8Lx0f7ri66AVj8uWmQzs
LhpI2rV5EDU7uPKjs7Txetb2PTHDLpmgwB6VlBpsIr/omzEeTGVAfT/zlSv4lyhh
QbpyEBH9ccZV9pJ3hdd8bVzizjhm/yn1Igqrpi7QUekCfu7U53PV4r1UBJP068bM
tOvWeNoySlsz6JPOrTSvOJSDIlx+tSI/WCeY7rWSC3+b48e9FIZZVnu4HKnqPi6k
DC4HWdM6ioTI1Q29BfmFA2I2OjyAKkLaZ/8XJj/wdW+l5KvGU2rHxryk010+h/u/
zFv73nG8ODVz0tN5qAnYuPYqctXkxnAYgTVcbqZVR9ERHluuIkIr3fhKRNZVBa4U
0qGN9BAjtf3J6m3KSPZHAC/fCeXvtJfuA0yeF/jWd3pGj9owV+G6qNDt260JwpX5
UuttSVF9t5Y0Sk5Gkeehi8swoMK9FhddJ2QbNNOBwfChbOUJPJ931CqntHBhV+Ti
xsPlhWpf
=fgO+
-----END PGP SIGNATURE-----
</signature_block>
</signature>
```

**Comment by herbdool** [comment]
Wow, this looks awesome so far. Thanks for the hard work!

**Comment by quicksketch** [comment]
Thanks @jlfranklin!!! I haven't yet had the time to review this. It _sounds_ great though! I'll get to this when I can.

**Comment by quicksketch** [comment]
I've been reviewing this one piece at a time. It's looks _really_ good. So far I'm stuck in project module changes, but I'll get to reviewing the code_sign modules soon.

**Comment by jlfranklin** [comment]
I rerolled this patch against 1.x this morning.

**Comment by jlfranklin** [comment]
I've opened #3714 to track updates to the Installer module to check the digital signatures.

**Comment by quicksketch** [comment]
Thanks @jlfranklin! With all the moving pieces between BackdropCMS.org, project module, and core, I don't think this is possible for 1.13.0, not your fault though, reviewing is where we got hung up here.

Let's continue pushing forward on the infrastructure changes and getting the feeds themselves updated with including signatures in the short-term, and shoot for having them validated by core in 1.14.0.

**Comment by klonos** [comment]
Which is the PR to be reviewed here?

**Comment by quicksketch** [comment]
The PR for this issue is https://github.com/backdrop/backdrop/pull/2449. I updated the summary to cross-link all the related issues as well.

**Comment by klonos** [comment]
Yeah, a single code_sign project, with the rest of the things being submodules seems more appropriate.

**Comment by jlfranklin** [comment]
Code Sign itself needs to be in core, because Installer calls it to verify the packages are legit.

**Comment by quicksketch** [comment]
> Code Sign itself needs to be in core, because Installer calls it to verify the packages are legit.

I'd like to get it on BackdropCMS.org first so it can sign the packages and update the XML feed. But as it doesn't look like we'll likely put it in 1.14, I'd like to have it available so we can install it separately. It also gives us a good opportunity to change it without any repercussions.

Though currently the only issue I've found is that the API for exposing signing profiles just has unnecessary hash sign prefixes: https://github.com/backdrop/backdrop/pull/2449/files#r317851694

**Comment by klonos** [comment]
...as also discussed during the last dev meeting, this should not have a release when moved to contrib.

**Comment by quicksketch** [comment]
Hash signs removed in https://github.com/backdrop-contrib/project/commit/b3ed711c55ec8c26f6656170f60f8964881c0c9f

The Contrib module (with sub-modules) for Code Sign is now pushed up to https://github.com/backdrop-contrib/code_sign, also with the hash signs removed.

**Comment by quicksketch** [comment]
I'm taking this out of the 1.15.0 milestone, as today is feature freeze and this is still not ready (mostly my fault I know). I'll continue to take this forward when I can but if there are any volunteers who are interested I'm happy to turn over the work.

**Comment by yorkshire-pudding** [comment]
RE: the discussion about sodium in the weekly meeting - https://www.youtube.com/watch?v=tQ-rkKxPkfw (timecode to be added)
I checked on one of my shared hosting sites, and the `phpinfo()` page has the following under Configure Command:
` '--with-sodium=shared,/usr' `
So may well be widely supported.

**Comment by avpaderno** [comment]
With the hosting provider I am using, I can enable the Sodium extension on any PHP version between 7.2 and 8.3. I am not sure how many hosting providers allow this or provide PHP with the extension already enabled.

This could be information provided by telemetry, but first Backdrop core should (optionally) use the Sodium extension. Differently, no site would enable that extension, and telemetry data would not provide useful data.

**Comment by yorkshire-pudding** [comment]
> With the hosting provider I am using, I can enable the Sodium extension on any PHP version between 7.2 and 8.3. I am not sure how many hosting providers allow this or provide PHP with the extension already enabled.
> 
> This could be information provided by telemetry, but first Backdrop core should (optionally) use the Sodium extension. Differently, no site would enable that extension, and telemetry data would not provide useful data.

I checked and while it was included in the build it wasn't enabled by default, but I was able to go and enable it like @avpaderno . I'm using cPanel

**Comment by avpaderno** [comment]
Instead of *this could be information provided by telemetry*, I should have said *how many sites have the Sodium extension enabled could be information provided by telemetry*. Differently, it seems I am talking of how many hosting providers allow to enable the Sodium extensions and how many hosting providers provide PHP with the Sodium extension enabled.

**Comment by indigoxela** [comment]
Re Sodium, and how widely it's a available out there I did some research in sites I maintain (quite a hosting provider zoo...). Hm, that's mixed.
In lots of PHP variants it seems to be automatically available/compiled (usually v 1.0.18), but several hosting providers seem to separate it out as indiviual module, that's off by default.
Notably if cPanel's in use it's off by default. I don't always have access to the hosting panel, so I can't tell for sure, but that's my finding.

Which means, we can't simply rely on Sodium being available (unlike, for example, json).

But I do belief, that digital signatures are more than just "nice to have". Especially, but not only with automatic updates.

**Comment by herbdool** [comment]
WP provides a compatibility library if a host doesn't include it. https://core.trac.wordpress.org/ticket/45806. It could be provided in core or in a contrib module. That's one option here. 

Another is to require libsodium if someone wants to install/update through the browser.

**Comment by avpaderno** [comment]
[paragonie/sodium_compat](paragonie/sodium_compat) is a pure-PHP polyfill. Instead of using that, if the matter is only including digital signatures on project packages, I would rather use the OpenSSL extension, which also handles certificates (something the Sodium extension does not handle).

I would use the Sodium extension when the intention is using Blake2, AEGIS-256, XCacha20-Poly1305, or other features only the Sodium extension has (such as the constant-time functions `sodium_bin2base64()`, `sodium_base642bin()`, `sodium_bin2hex()`, and `sodium_hex2bin()`).

**Comment by indigoxela** [comment]
This issue has come a long way. Starting with an idea to use MD5 (which nowadays probably nobody would consider anymore), over using OpenSSL, then GnuPG, then Sodium...

What are we really trying to achieve? And what would be the most reliable, backwards compatible approach, that doesn't require all module maintainers to make changes to all of their projects?

And we'd maybe also not want to maintain thousands (?) of lines of security related signature verification code in core? At least, to me this sounds like quite some maintenance burden.

Would it be acceptable to require signature verification for _automatic_ updates, but not for "regular" updates triggered by admins over UI (still, optional if available)?

Which base for signature verification would be the most promising for most environments? As of Backdrop 1.31 we will require PHP 7.1 at least, so we don't have to consider PHP 5 anymore in this discussion.

How do other projects do that, and which approach provides sufficient security for us, while not requiring too much from (shared) hosting, where probably the majority of Backdrop sites are?

**Comment by indigoxela** [comment]
Another thing to consider: while updates fetched from our server (`updates.backdropcms.org`) will probably be the norm, it's not the only possible source for project updates AFAIK.

**Comment by avpaderno** [comment]
Drupal seems to aim to use [The Update Framework (TUF)](https://theupdateframework.io/), which considers [three signature schemes](https://theupdateframework.github.io/specification/v1.0.33/#file-formats-keys):

- rsassa-pss-sha256 (RSA probabilistic signature scheme)
- ed25519
- ecdsa-sha2-nistp256

Of those, the first and the second can be implemented using the OpenSSL extension, although certificates do not seem involved in those schemes.  
The second could be implemented using the OpenSSL extension, but it would require PHP 8.4. It can be easily implemented using the Sodium extension, which would also offer constant-time functions for encoding/decoding keys (which should be used because they do not allow to gather information about the key basing on the time took to handle the keys, the same reason for which `hash_equal()` should be used to compare two hash values).

**Comment by avpaderno** [comment]
(I still have to understand whether the OpenSSL extension in PHP 8.4 is able to handle Ed25519/X25519/Ed448/X448 keys only with OpenSSL 3 or also with OpenSSL 1.1.)

**Comment by quicksketch** [comment]
Jumping back in here today as we're discussing this today as a [possible priority (again) for core development](https://backdropcms.org/news/backdrop-cms-core-priorities-we-want-your-input).

Things I'd like to see on this PR, starting from https://github.com/backdrop/backdrop/pull/2449:

- Convert this from a module that can be enabled/disabled to a core framework (i.e. under `/core/includes`) so it is always available as an API.
- Convert the implementations from being procedural with separate callback functions to using classes.

Right now the callbacks are specified by:
```
  $signers['hash'] = array(
    'title' => t('Hash'),
    'sign callback' => 'code_sign_hash_sign',
    'verify callback' => 'code_sign_hash_verify',
    'status callback' => 'code_sign_hash_status',
    'public keychain callback' => 'code_sign_hash_get_public_keychain',
    'private keychain callback' => 'code_sign_hash_get_private_keychain',
    'file' => backdrop_get_path('module', 'code_sign_hash') . '/hash.code_sign.inc',
  );
```

And this could/should be consolidated into a class that implements all the needed callbacks, more like this:
```
  $signers['hash'] = array(
    'title' => t('Hash'),
    'class' => 'CodeSignHash',
  );
```

**Comment by indigoxela** [comment]
Wow, this issue got revived! 🎉

Personal opinion: a lot of time has passed, since that issue has been opened. Nowadays signed packages are more than just nice to have. I support package signatures, regardless of automatic updates.

**Comment by yorkshire-pudding** [comment]
This should not derail anything, but assuming we do this on `backdrop-contrib` as well as core we need to think about whether this will work for bee when it does releases, or whether either:
- bee needs to move home so releases don't cause a problem
- bee needs some modification so that releases will work with signatures

**Comment by indigoxela** [comment]
Doesn't bee already do its own thing (building phar)? I guess, signatures could get added any time for bee, anyway. And isn't there an exemption for it on B-org? I'm sure (not) handling bee is feasible.

**Comment by yorkshire-pudding** [comment]
> Doesn't bee already do its own thing (building phar)? I guess, signatures could get added any time for bee, anyway. And isn't there an exemption for it on B-org? I'm sure (not) handling bee is feasible.

When bee does a release it goes to the packaging server and creates a release like other released modules except:

- Because there is no `.info` file a `PACKAGING_ERRORS.txt` file is added to the release which needs to be manually deleted
- I checkout the release tag locally and generate the `.phar` file which I then manually add to the release

See https://github.com/backdrop-contrib/bee/wiki/Release-Process for the full, currently manual process.

**Comment by izmeez** [comment]
This is a great idea. What about using SHA-256 instead?

**Comment by quicksketch** [comment]
> Because there is no .info file a PACKAGING_ERRORS.txt file is added to the release which needs to be manually deleted

Hm, that is interesting. When we actually create the signature, we need to store it somewhere. Some things (like the MD5 of the archive) don't need to be stored in the downloaded zip file because the hash of the zip file itself is the signature. But for other signatures, I think putting them in the .info file is the most reasonable place. The packager on backdropcms.org already adds lines to .info files, and adding a few more for the signature(s) would be logical.

For `bee` in general, I don't think it will be affected by the addition of signatures since you can't install bee through the module installer in the UI anyway. Signatures are only useful if they're actually checked.

Here's a breakdown of how it will work in practice:

1. When a release of a module/theme/layout is created on GitHub, it notifies backdropcms.org.
2. The packager (part of project module) modifies the .info file to add signatures. It generates the signatures with a private key stored on backdropcms.org.
3. When selecting a module or updating through the UI, `installer` module will get information about the public key from the JSON/XML feed that provides the module data.
4. `installer` module extracts the archive, reads the .info file to find the signature. It verifies that the public key from the feed can be used to validate the signature.
5. If all signatures do not match, perhaps do one of two things: 1) Prevent installing that module, the archive and downloaded code are deleted immediately, or 2) If configured, allow installing/updating the module anyway, warning the user that the signature was not valid.

So as far as `bee` is concerned, I don't think the process of updating/installing it will be affected, since it doesn't go through `installer` module in the first place.

What `bee` _will_ want to do however is to incorporate the same signature checking the `installer` module will use, so when updating projects through `bee`, they also get the same security benefit.

> bee needs to move home so releases don't cause a problem

I think we have a special project type "project placeholder" on BackdropCMS.org, but right now `bee` is a "module". Maybe we should switch it to a placeholder? That might fix the packaging errors since placeholders don't get packaged (I think).

**Comment by yorkshire-pudding** [comment]
> What bee will want to do however is to incorporate the same signature checking the installer module will use, so when updating projects through bee, they also get the same security benefit.

I have added an issue - https://github.com/backdrop-contrib/bee/issues/486

> I think we have a special project type "project placeholder" on BackdropCMS.org, but right now bee is a "module". Maybe we should switch it to a placeholder? That might fix the packaging errors since placeholders don't get packaged (I think).

From what I can tell in the project module, no checks are done before the package is created:
https://github.com/backdrop-contrib/project/blob/4b7428cb2d6547c3a2e87f3c8886d0f486a1f062/project_github/project_github.module#L29-L33

```php
    case 'release':
      if ($payload['action'] === 'published') {
        $errors = [];
        $project_node_result = NULL;
        $package_result = _project_github_create_package($errors, $payload);
```
I also tested with disabling "Enable releases" on the project node and creating a prerelease - the package was still created so I can't see that changing the project type on `b.org` will make any difference though logically, it might make more sense to be that, as long as we keep syncing from the README.

Perhaps the work for this functionality needs to exit creating the package if either project is placeholder or "enable releases" is disabled? It shouldn't be too difficult to replace what the packager does with a GitHub action; it might also spur me to add the `phar` creation into the GitHub action.

**Comment by quicksketch** [comment]
I filed a new PR at https://github.com/backdrop/backdrop/pull/5304 that reworks @jlfranklin's original module fairly substantially:

- Centralizes management of code sign profiles so each module does not need its own menu callbacks.
- Creates new `code_sign_sodium` signing engine suitable for signing/validating packages simply.
- Has the ability to bundle core public keys that need no individual profile configured: they're always available.
- Separates out the entire UI for creating signing profiles to a dedicated module: Code Sign UI.

What I sort of expect here is that we would hide most or all of these modules in the UI and enable/disable them in tandem with Installer module. It seems cleaner to keep these as separate modules but we could potentially bundle the code sign classes into Installer module directly.

Another concern/question: What to do about the UI. In the PR the UI is currently exposed, but a UI for creating code signing profiles in core is a bit of a stretch. In practice, only one site will ever be signing packages: backdropcms.org. Having a full UI for managing profiles in core seems excessive. BUT maintaining all the code in one place is a definite advantage. If we only put the API in core and put the UI in contrib, coordinating changes between them could be difficult. OTOH, maybe we should put the entire UI directly into [Project module](https://github.com/backdrop-contrib/project/), since its the only one that will be signing packages anyway.

In the mean time I think we can just review the overall approach while we also figure out what to do about the UI. I'd like to talk about this during the weekly dev meetings tomorrow.

**Comment by herbdool** [comment]
I think putting the UI into Project module makes the most sense. It might make coordination of changes more difficult but they're not going to be modules getting a lot of development once they're released.

**Comment by quicksketch** [comment]
Putting the UI into project module also reduces a level of abstraction. Right now we need to make a "Signing Profile" in Code Sign UI, then map/enable that to a "Project Code Sign Profile". Realistically, Project module is going to be the exclusive user of signing. I'll see about moving the UI into the Project module, replacing/updating what was added in https://github.com/backdrop-contrib/project/pull/30.

**Comment by indigoxela** [comment]
I think, there's consensus here to keep the part in core a slim as possible and let the project module hold the majority of code.

Some additional questions:

1. Will core require PHP 7.2 then? (Currently BACKDROP_MINIMUM_PHP is 7.1)
2. What happens if the private key changes - will all contrib packages get updated keys?
3. What happens with *core* updates if the private key changes?
4. What happens if the local (public) key is tainted and one attempts to update projects?
5. What does the UI look like if validation fails?
6. What happens if an older private key on B-org is lost?
7. What about additional package sources?

RE additional package sources... In theory that's abstracted in a way, as the `'project status url'` property could also get set in a module's info file. Someone *could* in theory provide their own infrastructure for updating custom projects. How does Silkscreen CMS handle that? Does this have to go?

Keep in mind that people often update with quite some delay. So an update may happen from a really, really old version.
Also Keep in mind that contrib projects can have very long-lived versions, if they require no updates.

**Comment by yorkshire-pudding** [comment]
> 2\. What happens if the private key changes - will all contrib packages get updated keys?
>
> Also Keep in mind that contrib projects can have very long-lived versions, if they require no updates.

I think this point will be a major consideration. On the introduction of this, will all contrib need a new release to be able to get installed?

If a project does not need a release as it works stably, will it still a new release every year or so, or will the rotation frequency of keys mean that older keys, while not used for new signings, will still validate for several years?

How will this be managed with projects that work without issue but have no current maintainer (or if the maintainer is not active or responsive)?

**Comment by indigoxela** [comment]
> I think this point will be a major consideration. On the introduction of this, will all contrib need a new release to be able to get installed?

Ouch, right, that wouldn't be feasible. 😬

What I'm struggling to understand: what `$data` will actually get signed / verified? And at which point? Before or after creating the zip file and uploading to GitHub release pages?

**Comment by quicksketch** [comment]
We discussed this today [during the weekly dev meeting](https://www.youtube.com/watch?v=sfYM3BABxaQ).

> What I'm struggling to understand: what `$data` will actually get signed / verified? And at which point? Before or after creating the zip file and uploading to GitHub release pages?

I've been struggling with answering that question as well.

@swampopus proposed the idea that we could use the `sha256` hash of the release archive as the `$data` that gets signed. And then instead of putting the signature in the archive, we put it _alongside_ the archive as an additional artifact on the release.

I believe that is a valid and secure way to do this. Because the only public keys included in core will be the ones that correspond to BackdropCMS.org's private keys. No one else would be able to generate and attach a signature that matches the available public keys.

But I think that brings me back to what I had originally thought: If the signatures can be separate and outside the archive, then we could put those keys directly into the project browser JSON feeds and update module XML feeds. We'd still need to store the signature somewhere though; that could be on the release node on BackdropCMS.org, which would prevent maintainers from accidentally deleting the signatures off their releases.

Assuming keeping the signature outside the archive is valid, what do folks think would be the better approach? Put the signatures on the GitHub releases, or put the signatures on the Backdrop CMS nodes?

**Comment by swampopus** [comment]
Just my 2 cents, but I like the idea of github being the official location for all the inner workings of Backdrop. So, I would vote against storing signatures purely within the backdropcms.org site as fields attached to release nodes.

As for moderators accidentally deleting the signature files from their releases-- I am not sure if this can work globally on all contribs, but I believe you could create a Github Action that acts once a day or hour, etc, and it could check for the presence of "module_name.sig" or whatever, and if it's missing, it could change the *name of the release* of the release to say "1.x-1.0.5 ! WARNING MISSING SIGNATURE"

Or I guess if it got to that point, where it knew the signature file was missing, it possibly just recreate the signature file itself automatically?

So, I know on a per-repository basis you can add github actions in this way.  But with the nature of github-contrib, I don't know if you can tell it to automatically inherit or copy the same action to every repo.  I just don't know enough about github to say for sure.

Or, if a project is missing the sig file on github, then the installer module has an extra confirmation screen saying "Potentially unsafe - missing signature"?

Of course, this might all be more complicated than just storing it as a field in the release node :)

**Comment by indigoxela** [comment]
> Just my 2 cents, but I like the idea of github being the official location for all the inner workings of Backdrop.

Whoah, that's a bit too enthusiastic about a free service, for which we have no SLA, no contract, no guarantee of any kind. 😆
My suggestion would rather be to now use GitHub with everything we can have for free, but also keep in mind, that things can change in the future. In other words, don't build crucial processes like package signing in a way that depends on something that we can't control.

But, seriously, I like the idea of a small, additional signature file _uploaded_ to GH release pages. That would also work for existing contrib projects. But the signing process should happen on B-org, of course.

One culprit, though: currently, if that signature would be 404, the check needs to get bypassed. It can absolutely be that a module has no such signature, yet. So a MiM attack has an easy way to bypass checking. 😜 Full protection can only exist, if it's a 100% check. 

@swampopus we should _not_ just rely on the existing GitHub sha256 - that would be so easy to spoof in a MiM.
Reliable signing requires some public/private key pair - which prevents spoofing.

**Comment by swampopus** [comment]
@indigoxela 
> Whoah, that's a bit too enthusiastic about a free service, for which we have no SLA, no contract, no guarantee of any kind.

What I meant is just that all of the code, issues, release notes, etc, for Backdrop core and all contrib modules is already linked directly to github. b-org gets all its info from github. It makes sense to me that _signatures_ are also stored on github, rather than in the nodes on b-org. I didn't mean the logic of generating the signature files have to come from github.  Just the _storage_ of the end result sig file.

> we should not just rely on the existing GitHub sha256 - that would be so easy to spoof in a MiM.
> Reliable signing requires some public/private key pair

Yep. So the idea is instead of:

``` (1) Download zip file -> (2) Compute hash of file -> (3) Use private key to create signature file ```

We'd skip a step and do this:

``` (1) Use github's existing hash of file -> (2) Use private key to create signature file ```

This has the benefit of probably being much faster than downloading every zip file of every contrib module release, as each sha256 string is just 64 characters, instead of however large the zip files are.

**Comment by indigoxela** [comment]
> I didn't mean the logic of generating the signature files have to come from github. Just the storage of the end result sig file.

Ahaaaa, then I got it totally wrong. Yes, I think it's fine to attach that sig file to the release page on GH. 😉 

> ... We'd skip a step and do this ...

That means, trusting and signing a download from somewhere else. 😉
My idea was to create the zip and sig file in one go *before* uploading to GitHub. Because downloading might introduce other flaws. I'm aware that this causes another problem - the one with old module releases, that aren't signed. Whatever the way to go will be - there's a lot to consider...

**Comment by swampopus** [comment]
I think I misunderstand how github provides releases.  I work on another project that has a github action that sets the release number in a file (based on the tag), then creates the zip file as an asset of the release.  This all happens inside of github.

But Backdrop doesn't do that?  b-org is responsible for creating the release zip file with the updated info file?  And then it sends it back to github?

**Comment by indigoxela** [comment]
> I think I misunderstand how github provides releases. I work on another project that has a github action...

I see. No, apparently the process in Backdrop is slightly different:

https://github.com/backdrop-contrib/project/tree/1.x-2.x/project_github

> Provide zip file "packages" on GitHub when a new release is created on GitHub. These packages are created by downloading the GitHub release zip file, modifying the contents (by scanning for .info files), and then attaching the modified package back to the release on GitHub.

The project_github submodule does download the source zip file, adapts it and uploads an *additional* zip file (info file updated, .github directory removed...).

I have to admit, I also got it wrong. I thought, that B-org does a _checkout_ on a tag - which isn't the case, either. 😉

**Comment by swampopus** [comment]
Well, should that process ever want to move into github, it's actually really easy.  You just add a file to a directory in your repo called .github/workflows.  Here's mine: https://github.com/flightpathacademics/flightpath/blob/flightpath-7.x/.github/workflows/release.yml

It even detects if the version string was entered wonky and changes the description to alert the user that they messed up the version string, as well as refusing to create the release zip file.  For the version, in my case I'm editing a value in a .inc file, but the logic would work just as well for a .info file.  

I know it puts the logic in github, but having a single yaml file is a lot easier (for me) than programming a webhook + php code to download a zip, modify a file, then re-upload.

**Comment by indigoxela** [comment]
> should that process ever want to move into github...

😆 I saw it coming... I mean, relying on GitHub Actions for package creation - what could possibly go wrong? Oh, wait... https://securitylabs.datadoghq.com/articles/case-for-github-actions-security/ 😈

**Comment by swampopus** [comment]
Heh.  Well, something to keep in our back pockets at any rate.

**Comment by yorkshire-pudding** [comment]
> Assuming keeping the signature outside the archive is valid, what do folks think would be the better approach? Put the signatures on the GitHub releases, or put the signatures on the Backdrop CMS nodes?

I think there could be advantages to doing both as depending on what is doing the downloading (installer/updater, backdrop or bee) then where the signature is could affect how many requests need to be made. If the information is being downloaded from `b.org` then it may save an additional request to GitHub. For downloads, bee gets all the information from GitHub, but when updates are finished then some data will come from `b.org`. Backdrop gets all the information from `b.org` and only gets the zip from GitHub.

**Comment by quicksketch** [comment]
Follow-up to @swampopus here:

> We'd skip a step and do this:
> `(1) Use github's existing hash of file -> (2) Use private key to create signature file`

I think this is a great idea for retroactively generating signatures for all existing contrib module releases. We could do that all in one big batch operation, and we wouldn't need to download or modify the existing release zip files. Since this would only need to run once, we don't need to worry about MiM attacks after that initial generation.

For _new_ releases, we probably would just `sha256` the entire release archive and then create a signature off of that, since the zip file will already be on backdropcms.org during the packaging. It will be easier and simpler to `sha256` the archive we just created, then append the signature(s) to the release on GitHub alongside the release archive.

I also like the idea of simply appending the signatures to the release so if we add or rotate the keys in the future, we can re-run the signing process again for all contrib and append another signature. Though that may mean down the line that we have several signature files attached to every release.

**Comment by avpaderno** [comment]
I apologize if this a too-ahead comment.
In the case multiple signatures are appended, is there a way to mark one or more signatures as obsolete? In the case a private/public key pair is compromised, I suppose signatures done with that pair should not be longer accepted, while old signatures should still be acceptable when the relative keys are not compromised.

**Comment by indigoxela** [comment]
> In the case a private/public key pair is compromised, I suppose signatures done with that pair should not be longer accepted

Definitely a good point.

**Comment by swampopus** [comment]
> Though that may mean down the line that we have several signature files attached to every release

I'd argue that for every release, there can only be one signature file (eg, module_name-1x-1.0.0.sig.txt).  And if the signatures need to get re-generated, then the old signature file just gets deleted and replaced.

Of course, if you have more than one signing key in use at the same time, this wouldn't work.  Like if you were slowly transitioning away from one kay pair and and to a new key pair before the old one expired.

In that case, could see that every release would have a _maximum_ of two signature files:
(1) module_name-1.x-1.0.0.sig.txt
(2) module_name-1.x-1.0.0.sig-latest.txt

And the logic there would be that the installer looks for the "latest" one first, and, failing to find it, looks for the default one.

I guess another option would be to have the "codename" or date (boring ;) of  the key pair as part of the signature filename, and the installer looks for whichever key filename matches the public key it has for verification.

Ex:
(1) module_name-1.x-1.0.0-happy-hippo.sig.txt
(2) module_name-1.x-1.0.0-ungrateful-ungulate.sig.txt

And let's say the backdrop site has "ungrateful-ungulate.public.key", so it knows to get that version of the sig file.

I know I might be making this way too complicated 😎

**Comment by avpaderno** [comment]
Truly, there is no need to have multiple signatures: Since a public key can only verify a signature created with the associated private key, one signature is sufficient.
This simplifies also the verification code, which would just loop over the available public keys until it finds the one for which the signature is validated.

**Comment by swampopus** [comment]
I guess it depends on where the public keys are stored. I was imagining they were installed _with the backdrop site_.  If that's the case, then you might not have the new public key to verify the signature.

But, if the public keys are checked/retrieved each time the user goes to install a project, then what you're describing would work.

**Comment by quicksketch** [comment]
Moving this back to needs work since I think this needs pretty significant refactoring:

- Remove the UI for creating signing profiles and move that to Project module (i.e. `project_code_sign`).
- Leave only the signing and verification code in core, which can probably exist as code in `/core/includes` rather than a hidden module.
- And eventually, add the validation into update and installer module.

What this means for the time being is that we can work on the infrastructure side of things, like working on Project module and backdropcms.org, instead of the core side. I feel like we've made a lot a great progress here in figuring out how this will actually work beyond just making a signing/verification library; but we're not going to get this completed in the next week to make it for 1.35.0.
